File: 04-scratchcards/app/Main.hs
{-# LANGUAGE RecordWildCards #-}

module Main (main) where

import           Control.Applicative          (Alternative (some))
import           Data.Char                    (isDigit)
import           Data.IntMap                  (IntMap, alter, elems, empty,
                                               fromList, (!))
import           Data.List                    (find)
import           Data.Maybe                   (fromJust)
import           Text.ParserCombinators.ReadP

data Card = Card
  { cid     :: Int
  , winNums :: [Int]
  , myNums  :: [Int]
  } deriving (Eq, Show)

winningNums :: [Int]  -- winning nums
            -> [Int]  -- my nums
            -> [Int]
winningNums = filter . flip elem

matches :: Card -> Int
matches Card{..} = length $ winningNums winNums myNums

points :: Int  -- winning num count
       -> Int
points 0 = 0
points 1 = 1
points n = 2 * points (n - 1)

cardPoints :: Card -> Int
cardPoints = points . matches

-- parsers

card :: ReadP Card
card = do _ <- string "Card"
          skipSpaces
          cid <- int
          _ <- char ':'
          skipSpaces
          wins <- sepBy1 int skipSpaces
          _ <- string " | "
          skipSpaces
          my <- sepBy1 int skipSpaces
          return $ Card cid wins my
  where int = read <$> munch1 isDigit

parseCard :: String -> Card
parseCard = fst . fromJust . find (null . snd) . readP_to_S card


-- main :: IO ()
-- main = do input <- readFile "input.txt"
--           let lined = lines input
--           let cards = map parseCard lined
--           let points = map cardPoints cards
--           print $ sum points

-- part 2

-- which cards do I won after this one card?
-- may include non-existing cards
type CardID = Int
cardsWon :: Card -> [CardID]
cardsWon c@Card{..} = [cid + 1 .. cid + matches c]

type Pile = IntMap CardID

startingPile :: Pile
startingPile = fromList $ map (, 1) [1..219]

totalCards :: [Card] -> Pile
totalCards = foldl accum startingPile

accum :: Pile -> Card -> Pile
accum m c@Card{..} = foldl f m (cardsWon c)
  where f :: Pile -> CardID -> Pile
        f m i = alter (maybe (Just n) (Just . (+n))) i m
        n = m ! cid

main :: IO ()
main = do input <- readFile "input.txt"
          let lined = lines input
          let cards = map parseCard lined
          let total = totalCards cards
          let occurrences = elems total
          print $ sum occurrences

File: 04-scratchcards/main.py
# pyright: strict

from dataclasses import dataclass
from typing import Dict, List


@dataclass
class Card:
    cid: int
    win_nums: List[int]
    my_nums: List[int]


def parse_card(line: str) -> Card:
    [label, rest] = line.split(': ')
    [_, cid_str] = label.split()
    cid = int(cid_str)
    [win, my] = rest.split(' | ')
    wins = map(int, win.split())
    mine = map(int, my.split())
    return Card(cid=cid, win_nums=list(wins), my_nums=list(mine))


# <Card 1: 41 48 83 86 17 | 83 86  6 31 17  9 48 53> -> 8
def card_point(card: Card) -> int:
    total_wins = len(list(filter(lambda n: n in card.win_nums, card.my_nums)))
    return 0 if total_wins == 0 else 2 ** (total_wins - 1)


def main_():
    handle = open('input.txt', 'r')
    lines = handle.read().splitlines()
    handle.close()
    cards = map(parse_card, lines)
    points = map(card_point, cards)
    total = sum(points)
    print(total)


# part 2


# card id and its number
Pile = Dict[int, int]

# id to card
Cards = Dict[int, Card]


def cards_won(card: Card) -> List[int]:
    total_wins = len(list(filter(lambda n: n in card.win_nums, card.my_nums)))
    return list(range(card.cid + 1, card.cid + total_wins + 1))


# welp, i tried to be more functional but this is not haskell
# and everything python offers about dict are in place mutations
def update_sequantially(pile: Pile, cards: Cards):
    ids = pile.keys()
    for cid in ids:
        number = pile[cid]
        card = cards[cid]
        wins = cards_won(card)
        for w in wins: pile[w] += number


def main():
    handle = open('input.txt', 'r')
    lines = handle.read().splitlines()
    handle.close()

    pile = { c.cid: 1 for c in map(parse_card, lines) }
    cards = { c.cid: c for c in map(parse_card, lines) }
    update_sequantially(pile, cards)
    print(sum(pile.values()))


if __name__ == '__main__':
    main()
